spring-cloud-zookeeper-discovery/src/main/java/org/springframework/cloud/zookeeper/discovery/configclient/ZookeeperConfigServerBootstrapper.java

```diff
@@ -168,22 +168,22 @@ public List<ServiceInstance> apply(String serviceId, Binder binder, BindHandler
 				return Collections.emptyList();
 			}
 
-			ZookeeperProperties properties = binder.bind(ZookeeperProperties.PREFIX, Bindable.of(ZookeeperProperties.class))
-					.orElse(new ZookeeperProperties());
-			RetryPolicy retryPolicy = new ExponentialBackoffRetry(properties.getBaseSleepTimeMs(), properties.getMaxRetries(),
-					properties.getMaxSleepMs());
+			ZookeeperProperties properties = context.getOrElse(ZookeeperProperties.class, binder.bind(ZookeeperProperties.PREFIX, Bindable.of(ZookeeperProperties.class))
+					.orElse(new ZookeeperProperties()));
+			RetryPolicy retryPolicy = context.getOrElse(RetryPolicy.class, new ExponentialBackoffRetry(properties.getBaseSleepTimeMs(), properties.getMaxRetries(),
+					properties.getMaxSleepMs()));
 			try {
-				CuratorFramework curatorFramework = CuratorFactory.curatorFramework(properties, retryPolicy, Stream::of,
-						() -> null, () -> null);
-				InstanceSerializer<ZookeeperInstance> serializer = new JsonInstanceSerializer<>(ZookeeperInstance.class);
-				ZookeeperDiscoveryProperties discoveryProperties = binder.bind(ZookeeperDiscoveryProperties.PREFIX, Bindable
+				CuratorFramework curatorFramework = context.getOrElse(CuratorFramework.class, CuratorFactory.curatorFramework(properties, retryPolicy, Stream::of,
```

Comment:
Would we ever reach the `orElse` here? I mean even before the lambda gets registered, that will later create this `ZookeeperFunction`, the code has already called `CuratorFactory.registerCurator`. (I may be missing some use-case though.)

Reply:
I was probably being over cautious here, there is no harm though.

Reply:
Apart from being a bit confusing to readers (who may expect this line to be creating their `CuratorFramework` instance), won't this `getOrElse` code always end up running the `CuratorFactory.curatorFramework(...)` call? If so, that would always [build another `CuratorFramework` instance, start it, and block while it connects (or fails to do so)](https://github.com/spring-cloud/spring-cloud-zookeeper/blob/4.0.x/spring-cloud-zookeeper-core/src/main/java/org/springframework/cloud/zookeeper/CuratorFactory.java#L44-L81), only to then throw it away. When this instance can't connect, it may also write a considerable amount of misleading logs, that could become red herrings when debugging some issue.

Reply:
It would only call `CuratorFactory.curatorFramework` if there was not one already present in the `BootstrapContext`.  

I don't see how it would start 2...

Reply:
I may be missing something, but my thinking was that [the `other` parameter of this `getOrElse` call](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/BootstrapContext.html#getOrElse(java.lang.Class,T)) has to be evaluated first by the JVM, meaning that the `CuratorFramework` instance gets built & started, then the JVM passes it into `getOrElse` as the second parameter, and when `getOrElse` gets run it will simply ignore that parameter when it realizes that there's already a `CuratorFramework` in the `BootstrapContext`.